generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -0,0 +1,167 @@
+package chipyard.iobinders
+
+import chisel3._
+
+import freechips.rocketchip.config.{Field, Config, Parameters}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import freechips.rocketchip.devices.debug._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.system._
+import freechips.rocketchip.util._
+
+import sifive.blocks.devices.gpio._
+import sifive.blocks.devices.uart._
+
+import testchipip._
+import icenet._
+import tracegen.{HasTraceGenTilesModuleImp}
+
+import scala.reflect.{ClassTag, classTag}
+
+// System for instantiating binders based
+// on the scala type of the Target (_not_ its IO). This avoids needing to
+// duplicate harnesses (essentially test harnesses) for each target.
+//
+// You could just as well create a custom harness module that instantiates
+// bridges explicitly, or add methods to
+// your target traits that instantiate the bridge there (i.e., akin to
+// SimAXI4Mem). Since cake traits live in Rocket Chip it was easiest to match
+// on the types rather than change trait code.
+
+
+
+// A map of partial functions that match on the type the DUT (_not_ it's
+// IO) to generate an appropriate bridge. You can add your own binder by adding
+// a new (key, fn) pair. You should override existing pairs in this map when
+// using a custom IOBinder
+
+// Since we also want to compose this structure like the existing config system,
+// use the scala string representation of the matched trait as a key
+
+case object IOBinders extends Field[Map[String, (Clock, Bool, Bool, Any) => Seq[Any]]](Map())
+
+
+// This macro overrides previous matches on some Top mixin. This is useful for
+// binders which modify IO, since those typically cannot be composed
+class RegisterIOBinder[T](fn: => (Clock, Bool, Bool, T) => Seq[Any])(implicit tag: ClassTag[T]) extends Config((site, here, up) => {
+  case IOBinders => up(IOBinders, site) + (tag.runtimeClass.toString ->
+      ((clock: Clock, reset: Bool, success: Bool, t: Any) => {
+        t match {
+          case top: T => fn(clock, reset, success, top)
+          case _ => Nil
+        }
+      })
+  )
+})
+
+// This macro composes with previous matches on some Top mixin. This is useful for
+// annotation-like binders, since those can typically be composed
+class RegisterBinder[T](fn: => (Clock, Bool, Bool, T) => Seq[Any])(implicit tag: ClassTag[T]) extends Config((site, here, up) => {
+  case IOBinders => up(IOBinders, site) + (tag.runtimeClass.toString ->
+      ((clock: Clock, reset: Bool, success: Bool, t: Any) => {
+        t match {
+          case top: T => fn(clock, reset, success, top) ++
```

Comment:
Unless i'm mistaken the the function associated with this Binder (lets call it _A_) will be invoked first, followed by the older one (_B_). This means that hardware generated by _A_ maybe overridden by hardware generated in _B_, which i'm not sure is the intuitive result.

Reply:
What is the intuitive result when composing 2 Binders that attach to the same IO? My thought was that the `RegisterBinder` macro would result in composition, while the `RegisterIOBinder` macro would override.

Reply:
I'd expect that anything you put in an LHS config instance should take priority over anything in an RHS config instance. Which is to say last connect semantics should favor connections made in an LHS config, which does not appear to be the case here (?).

Reply:
Oh I see, yeah I'll switch the ordering then.